src/Microsoft.AspNetCore.SystemWebAdapters.CoreServices/HttpRequestAdapterFeature.cs

```diff
@@ -0,0 +1,196 @@
+    }
+
+    Stream IHttpRequestAdapterFeature.GetBufferlessInputStream()
+    {
+        if (Mode is ReadEntityBodyMode.Bufferless or ReadEntityBodyMode.None)
+        {
+            Mode = ReadEntityBodyMode.Bufferless;
+            return GetBody();
+        }
+
+        throw new InvalidOperationException("GetBufferlessInputStream cannot be called after other stream access");
+    }
+
+    Stream IHttpRequestAdapterFeature.InputStream
+    {
+        get
+        {
+            if (Mode is ReadEntityBodyMode.Classic && _bufferedStream is not null)
+            {
+                return _bufferedStream;
+            }
+
+            throw new InvalidOperationException("InputStream must be prebuffered");
+        }
+    }
+
+    async Task<Stream> IHttpRequestAdapterFeature.GetInputStreamAsync(CancellationToken token)
+    {
+        await BufferInputStreamAsync(token);
+        return GetBody();
+    }
+
+    public async Task BufferInputStreamAsync(CancellationToken token)
+    {
+        if (Mode is ReadEntityBodyMode.Classic)
+        {
+            return;
+        }
+
+        if (Mode is not ReadEntityBodyMode.None)
+        {
+            throw new InvalidOperationException("InputStream cannot be called after other stream access");
+        }
+
+        var stream = GetBufferedInputStream();
+        await stream.DrainAsync(token);
+        stream.Position = 0;
+
+        Mode = ReadEntityBodyMode.Classic;
+    }
+
+    public void Dispose() => _bufferedStream?.Dispose();
+
+    string IHttpRequestFeature.Protocol
+    {
+        get => _other.Protocol;
+        set => _other.Protocol = value;
+    }
+
+    string IHttpRequestFeature.Scheme
+    {
+        get => _other.Scheme;
+        set => _other.Scheme = value;
+    }
+
+    string IHttpRequestFeature.Method
+    {
+        get => _other.Method;
+        set => _other.Method = value;
+    }
+
+    string IHttpRequestFeature.PathBase
+    {
+        get => _other.PathBase;
+        set => _other.PathBase = value;
+    }
+
+    string IHttpRequestFeature.Path
+    {
+        get => _other.Path;
+        set => _other.Path = value;
+    }
+
+    string IHttpRequestFeature.QueryString
+    {
+        get => _other.QueryString;
+        set => _other.QueryString = value;
+    }
+
+    string IHttpRequestFeature.RawTarget
+    {
+        get => _other.RawTarget;
+        set => _other.RawTarget = value;
+    }
+
+    IHeaderDictionary IHttpRequestFeature.Headers
+    {
+        get => _other.Headers;
+        set => _other.Headers = value;
+    }
+
+    Stream IHttpRequestFeature.Body
+    {
+        get
+        {
+            var body = GetBody();
+
+            if (Mode is ReadEntityBodyMode.None)
+            {
+                Mode = body.CanSeek ? ReadEntityBodyMode.Buffered : ReadEntityBodyMode.Bufferless;
+            }
+
+            return body;
+        }
+        set => _other.Body = value;
```

Comment:
Should Mode be updated if the stream is replaced?

Reply:
_bufferedStream as well?

Reply:
https://github.com/dotnet/systemweb-adapters/pull/173

Reply:
I had thought about it, but wasn't sure how necessary it would be. Pushed a PR to reset in that case